lib/ukbitops: Rework bit find ops for consistency + major code cleanup

andreittr opened the pull request:
### Description of changes

This change reworks the operations handling finding the first/last set bit in a word (`uk_ffs(l)` & `uk_fls(l)`) to be consistent with their similarly named libc & compiler builtin equivalents.
New operations based off builtins: `uk_ffs(l)`, `uk_clz(l)`, `uk_ctz(l)`.
New operations `uk_mssb(l)`, replacing previous `uk_fls(l)`, & `uk_lssb(l)`, alias for `uk_ctz(l)`.

Calling code has been updated to call the function they mean.
Arch-specific code has been removed, as all operations can be implemented in terms of compiler builtins.

Building on top of this, the headers of lib/ukbitops are reorganized, each with a well-defined scope.

Conflicts with the following PR; either need patching after the other is merged first:
- https://github.com/unikraft/unikraft/pull/1576

### Prerequisite checklist


 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.


### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]


### Additional configuration

N/A

andreittr committed:
lib/ukbitops: Remove unneeded exportsyms.uk

This change deletes exportsyms.uk from ukbitops, as all functions are
provided as static inlines in headers, thus never requiring linking.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
lib/ukbitops: Rework bit find ops for consistency

This change reworks the operations handling finding the first/last set
bit in a word (uk_ffs(l) & uk_fls(l)) to be consistent with their
similarly named libc & compiler builtin equivalents.
New operations based off builtins: uk_ffs(l), uk_clz(l), uk_ctz(l).
New operations uk_mssb(l), replacing previous uk_fls(l), & uk_lssb(l),
alias for uk_ctz(l).

Calling code has been updated to call the function they mean.
Arch-specific code has been removed, as all operations can be
implemented in terms of compiler builtins.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
drivers/virtio: Use bitops instead of bitmap API

This change removes the use of `uk_test_bit` from virtio MMIO code, as
that expects a bitmap of arbitrary length, in favor of the core ukbitops
API.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
lib/ukintctlr: Use atomic ops instead of bitmap

This change removes the use of __uk_test_and_set_bit (part of the bitmap
API) in favor of using atomic ops explicitly.
Config.uk dependencies have been updated to reflect reality.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
plat/kvm/x86: Use atomic ops instead of bitmap API

This change removes the use of __uk_test_and_clear_bit (part of the
bitmap API) in favor of using atomic ops explicitly in x86 tscclock
code. Config.uk has been updated to reflect this.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
plat/common/arm: Use atomic ops instead of bitmap

This change removes the use of __uk_test_and_clear_bit (part of the
bitmap API) in favor of using atomic ops explicitly in arm time code.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
plat/xen: Add missing dependency on ukbitops

This change adds a missing dependency to libukbitops in the Config.uk
file of xen. In addition, the use of the bitmap API in events.c has been
made explicit by the inclusion of bitmap.h.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
lib/ukbitops: Move bitmap funcs out of bitops.h

This change moves all bitmap-related functionality out of the main
bitops.h into its own header.

Checkpatch-Ignore: PARENTHESIS_ALIGNMENT
Checkpatch-Ignore: VOLATILE
Checkpatch-Ignore: SPACING
Checkpatch-Ignore: UNNECESSARY_PARENTHESES
Checkpatch-Ignore: MACRO_ARG_REUSE
Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
lib/ukbitops: Remove bitcount.h from bitops.h

This change removes the convenience inclusion of bitcount.h in the main
bitops.h header. Users of the header have been updated to include
bitcount.h explicitly.
In addition, the uk_hweight* macros have been removed, as they only
served as aliases for bitcount.h ops.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
lib/ukbitops: Remove bitscan.h from bitops.h

This change removes the convenience inclusion of bitscan.h in the main
bitops.h header. Users of the header have been updated to include
bitscan.h explicitly.
In addition, the unused uk_get_count_order inline has been removed from
bitops.h.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
lib/ukbitops: Move bitcount.h under lib namespace

This change moves bitcount.h under bitops/ in the default include path,
as is convention for secondary library headers.
Users of bitcount.h have been updated.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
lib/ukbitops: Move bitmap.h under lib namespace

This change moves bitmap.h under bitops/ in the default include path,
as is convention for secondary library headers.
Users of bitmap.h have been updated.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
lib/ukbitops: Add missing dependency to callers

This change adds a missing dependency to libukbitops to all core drivers
and libraries that include any of its headers. This makes the Config.uk
dependency tree better reflect reality.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr commented:
Pushed a major general code cleanup of `libukbitops` in addition to some tweaks to the original rework.

andreittr commented:
Pushed minor fix.

michpappas reviewed (approved):
Reviewed-by: Michalis Pappas <michalis@unikraft.io>

mogasergiu commented:
Why did we switch to using non-relaxed atomic operations on things that would normally be per-CPU, such as  `sched_have_pending_events`?

andreittr committed:
lib/ukatomic: Add relaxed memory ordering ops

This change adds variants of uk atomic ops that use a relaxed memory
ordering, as opposed to the total ordering that the default ops use.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr commented:
Pushed update addressing last comment; previous version had unintentionally tightened memory order requirements.

mogasergiu reviewed (approved):
Thanks! Although we get to keep the relaxed ordering again, I still don't understand why these had atomic semantics in the first place. As far as I can tell they would never be executed in parallel except for cases where said variables would be per-CPU (e.g., the sched pending events variable). (also why do none of the macros in that header use () around parameters :thinking: )

Anyways, maybe not worth pondering on this right now. Worth revisiting when we do SMP.

Reviewed-by: Sergiu Moga <sergiu@unikraft.io>

michpappas commented:
Approved-by: Michalis Pappas <michalis@unikraft.io>
